File: services/mobile/lib/src/lyrics/lrc_parser.dart
import 'package:inori_music/src/lyrics/lyric_line.dart';

/// Parses LRC-format lyrics into a list of [LyricLine]s sorted by timestamp.
///
/// Handles standard [MM:SS.xx] and [MM:SS.xxx] timestamps.
/// Also parses enhanced (word-level karaoke) LRC inline tags of the form
/// `<mm:ss.xx>word`, producing per-word timing fragments in [LyricLine.words].
/// Lines without inline tags keep [LyricLine.words] as null.
/// Metadata tags like [ti:], [ar:], [al:] are ignored.
class LrcParser {
  static final _lineTagRegex =
      RegExp(r'^\[(\d{2}):(\d{2})[.:](\d{2,3})\](.*)$');
  static final _wordTagRegex = RegExp(r'<(\d{2}):(\d{2})[.:](\d{2,3})>');

  static List<LyricLine> parse(String content) {
    final lines = <LyricLine>[];
    for (final raw in content.split('\n')) {
      final line = raw.trim();
      final m = _lineTagRegex.firstMatch(line);
      if (m == null) continue;
      final timestamp = _parseDuration(m.group(1)!, m.group(2)!, m.group(3)!);
      final rawText = (m.group(4) ?? '').trim();
      if (rawText.isEmpty) continue; // skip metadata-only tags

      final words = _parseWords(rawText, timestamp);
      final text = words != null ? words.map((w) => w.text).join().trim() : rawText;
      if (text.isEmpty) continue;
      lines.add(LyricLine(timestamp: timestamp, text: text, words: words));
    }
    lines.sort((a, b) => a.timestamp.compareTo(b.timestamp));
    return lines;
  }

  /// Splits a line's raw text on inline <mm:ss.xx> tags into word fragments.
  /// Returns null when the line has no inline tags (plain line-level lyrics).
  /// Text preceding the first inline tag, if any, is kept as a fragment
  /// timed at the line's own [lineTimestamp].
  static List<LyricWord>? _parseWords(String rawText, Duration lineTimestamp) {
    final matches = _wordTagRegex.allMatches(rawText).toList();
    if (matches.isEmpty) return null;
    final words = <LyricWord>[];
    if (matches.first.start > 0) {
      words.add(LyricWord(
        offset: lineTimestamp,
        text: rawText.substring(0, matches.first.start),
      ));
    }
    for (var i = 0; i < matches.length; i++) {
      final m = matches[i];
      final offset = _parseDuration(m.group(1)!, m.group(2)!, m.group(3)!);
      final start = m.end;
      final end = i + 1 < matches.length ? matches[i + 1].start : rawText.length;
      if (end <= start) continue;
      words.add(LyricWord(offset: offset, text: rawText.substring(start, end)));
    }
    return words.isEmpty ? null : words;
  }

  static Duration _parseDuration(String minStr, String secStr, String csStr) {
    final min = int.parse(minStr);
    final sec = int.parse(secStr);
    // normalise to milliseconds: 2 digits = centiseconds, 3 = milliseconds
    final ms = csStr.length == 2 ? int.parse(csStr) * 10 : int.parse(csStr);
    return Duration(minutes: min, seconds: sec, milliseconds: ms);
  }
}

File: services/web/lib/lyrics/lrcParser.ts
import type { LyricLine, LyricWord } from "./lyricLine";

/**
 * Parses LRC-format lyrics into a list of LyricLines sorted by timestamp.
 *
 * Handles standard [MM:SS.xx] and [MM:SS.xxx] timestamps.
 * Also parses enhanced (word-level karaoke) LRC inline tags of the form
 * `<mm:ss.xx>word`, producing per-word timing fragments in LyricLine.words.
 * Lines without inline tags keep LyricLine.words as undefined.
 * Metadata tags like [ti:], [ar:], [al:] are ignored.
 */
const LINE_TAG_REGEX = /^\[(\d{2}):(\d{2})[.:](\d{2,3})\](.*)$/;
const WORD_TAG_REGEX = /<(\d{2}):(\d{2})[.:](\d{2,3})>/g;

export function parseLrc(content: string): LyricLine[] {
  const lines: LyricLine[] = [];
  for (const raw of content.split("\n")) {
    const line = raw.trim();
    const m = LINE_TAG_REGEX.exec(line);
    if (!m) continue;
    const timestampMs = parseTimestamp(m[1], m[2], m[3]);
    const rawText = (m[4] ?? "").trim();
    if (rawText.length === 0) continue; // skip metadata-only tags

    const words = parseWords(rawText, timestampMs);
    const text = words ? words.map((w) => w.text).join("").trim() : rawText;
    if (text.length === 0) continue;
    lines.push({ timestampMs, text, words });
  }
  lines.sort((a, b) => a.timestampMs - b.timestampMs);
  return lines;
}

/**
 * Splits a line's raw text on inline <mm:ss.xx> tags into word fragments.
 * Returns undefined when the line has no inline tags (plain line-level lyrics).
 * Text preceding the first inline tag, if any, is kept as a fragment
 * timed at the line's own lineTimestampMs.
 */
function parseWords(rawText: string, lineTimestampMs: number): LyricWord[] | undefined {
  const matches = [...rawText.matchAll(WORD_TAG_REGEX)];
  if (matches.length === 0) return undefined;
  const words: LyricWord[] = [];
  const first = matches[0];
  if (first.index! > 0) {
    words.push({ offsetMs: lineTimestampMs, text: rawText.slice(0, first.index) });
  }
  for (let i = 0; i < matches.length; i++) {
    const m = matches[i];
    const offsetMs = parseTimestamp(m[1], m[2], m[3]);
    const start = m.index! + m[0].length;
    const end = i + 1 < matches.length ? matches[i + 1].index! : rawText.length;
    if (end <= start) continue;
    words.push({ offsetMs, text: rawText.slice(start, end) });
  }
  return words.length === 0 ? undefined : words;
}

function parseTimestamp(minStr: string, secStr: string, csStr: string): number {
  const min = Number.parseInt(minStr, 10);
  const sec = Number.parseInt(secStr, 10);
  // normalise to milliseconds: 2 digits = centiseconds, 3 = milliseconds
  const ms = csStr.length === 2 ? Number.parseInt(csStr, 10) * 10 : Number.parseInt(csStr, 10);
  return min * 60_000 + sec * 1000 + ms;
}
